Generated configuration 9.2.2

LightGuard opened the pull request:
Reworking the earlier Pull Request for configuration properties.

gitgabrio reviewed (commented):
Thanks @LightGuard for the PR.
I have the impression there are some properties that need to be better defined, and some that are completely missing (I add them in a comment).

gitgabrio commented:
Thanks @LightGuard 
IINW, the following part/properties are missing

Engine names

decisions
persistence
predictions
processes
rules
sw

Global properties

kogito.generate.rest -> (boolean) enable/disable global rest endpoint generation (default true)
kogito.generate.di -> (boolean) dependency injection is available and enabled (default true)
kogito.codegen.ignoreHiddenFiles -> (boolean) whether Kogito Codegen should ignore hidden files (default true)
kogito.rest.resource.type -> the type of generated rest (used only by processes); possible values: reactive; (default is empty)
kogito.addon.cloudevents.kafka.kogito_incoming_stream -> 
kogito.addon.cloudevents.kafka.kogito_outgoing_stream ->

engine-specific properties

kogito.codegen.(engine_name) -> (boolean) enable/disable engine code generation (default true)
kogito.generate.rest.(engine_name) -> (boolean) enable/disable engine rest endpoint generation (default true)

decisions-specific properties

kogito.decisions.stronglytyped -> (boolean) generate java classes to support strongly typed input (default false)
kogito.decisions.validation -> enable(disable model validation); possible values: ENABLED, DISABLED, IGNORE; (default ENABLED)
kogito.addon.tracing.decision.kafka.bootstrapAddress → (string) kafka bootstrap server address
kogito.addon.tracing.decision.asyncEnabled -> (boolean) default true

persistence-specific properties

kogito.persistence.data-index.proto.generation -> (boolean) (default true)
kogito.persistence.proto.marshaller -> (boolean) (default true)
kogito.persistence.type -> possible values: (string) filesystem, infinispan, mongodb, postgresql, kafka, jdbc; (default infinispan)

rules-specific properties

kogito.rules.(rule_unit_name).event-processing-type -> (string) possible values: CLOUD, STREAM; (default is null)
kogito.rules.(rule_unit_name).clock-type -> (string) possible values: PSEUDO, REALTIME; (default is null)
kogito.rules.(rule_unit_name).sessions-pool -> size of the session pool; (default is null)

workflow-specific properties

kogito.workflow.version-strategy -> if "project", requires project GAV;  (default is workflow)

serverless-specific properties

kogito.sw.operationIdStrategy -> (string)  strategy for generating the configuration key of open API specifications; possible values: file_name (uses the last element of the spec uri), full_uri (uses the full path of the uri) spec_title (uses the spec title), function_name (uses the function name),
kogito.sw.functions.port -> ?

LightGuard commented:
Where are all of these properties defined? I looked for every occurrence of @ConfigItem and @ConfigGroup and did not find these in the codebase. Are they defined somewhere else? We're also not adding SW to this as SW is not part of BAMOE.

LightGuard commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+|boolean
+|`true`
+
+a| kie.flyway.modules.`<module name>`.enabled
```
Do we have a decent definition of a KIE module somewhere? I'm not sure if it is actually defined anywhere in the BAMOE docs.

gitgabrio replied:
@LightGuard 
I perfectly agree with you, I have the impression it is sort of left-over from pre-kogito-times.
@baldimir ?

pefernan replied:
Hi, this is to kie-flyway terminology, not to pre-kogito-times. Each maven artifact (jar) containing a`kie-flyway.properties` is considered a module. The properties configured in that properties look like:

```
module.name=data-index
module.locations.postgresql=...
module.locations.h2=...
```

This config is used by kie-flyway to configure which modules can be skipped from the initial migration, for example:

```
kie.flyway."data-index".enabled=false
```

LightGuard commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+a|false
+
+a|kogito.data-index.url
+a|URL the data index will use.
```
I could take a stab at http and ws, but no clue about data-index.url

pefernan replied:
This properties are used to configure data-index urls, for different purposes. Both `kogito.data-index.url` & `kogito.dataindex.http.url` are the same, and actually I'm tempted to say that we should consolidate them into a single property.

I think (based on the context where I found those props)`kogito.dataindex.http.url` & `kogito.dataindex.ws.url` were intended to be used to configure the data-index clients in any of the apps/services in the kogito umbrella (for example consoles or the kogito runtime if using the process-svg-addon). 
* `kogito.dataindex.http.url` is mainly used to connect the runtime (with svg addon) to data-index
*  `kogito.dataindex.ws.url`  I seen it in the examples but I have not clue about how it is used.

LightGuard commented:
@gitgabrio please take a look now. I've added the properties you've mentioned. I also have grouped the configuration properties and added some copy to help explain things.

gitgabrio commented:
> Where are all of these properties defined? I looked for every occurrence of @configitem and @ConfigGroup and did not find these in the codebase. Are they defined somewhere else? We're also not adding SW to this as SW is not part of BAMOE.

I found them in different places: e.g.

`kogito.codegen.(engine_name) -> org.kie.kogito.codegen.api.Generator`
`kogito.generate.rest(.engine_name) -> org.drools.codegen.common.DroolsModelBuildContext`

Those are not defined as `@ConfigGroup` (or similar, that are framework-specific) but as static final string (sometimes built during execution)

gitgabrio commented:
> @gitgabrio please take a look now. I've added the properties you've mentioned. I also have grouped the configuration properties and added some copy to help explain things.

Great stuff, @LightGuard 

Alas, there are still some missing bits for which, unfortunately, I have no idea:

org.kie.kogito.addons.knative.eventing.sink.api-version -> possible values
org.kie.kogito.addons.knative.eventing.sink.kind -> possible values
kogito.grpc.stream.timeout -> time unit
kogito.devservices.shared -> quarkus only ?
kogito.security.auth.impersonation.allowed-for-roles -> where are roles defined ?
multiple kogito.data-index urls

kogito.devservices.* -> are those available only for quarkus ?
there are some kogito.quarkus.events.* and kogito.events.* properties -> when are those available ?


@baldimir @porcelli @yesamer @pefernan 

Maybe you can help here

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+a|Kubernetes API Version of the given Knative Sink.
+This sink is used to configure the "sink" property in the generated Knative SinkBinding for this Kogito service.
+|string
+|`eventing.knative.dev/v1`
```
hmm not sure about this one tbh

LightGuard replied:
I'm removing (commenting out so we still have it for later if needed) Knative as it isn't supported as part of BAMOE

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+|
+
+a|kogito.devservices.shared
+a|Toggle to enable sharing of Quarkus Dev Services.
```
@gitgabrio yes, this is quarkus specific. In my understading it is used to share the data-index started by the Kogito DevServices between different kogito apps (in dev mode).  I think the whole use of the Data Index started by kogito dev services is not recommended for our setup and we are enforcing using the right data-index add-on.

pefernan replied:
I think the full name of this propertry is `quarkus.kogito.devservices.shared`

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+|`1000`
+
+a|kogito.devservices.enabled
+a|Toggle to enable Kogito Dev Services.
```
This is to enable or disable the Kogito Dev services in dev mode, which is start an In-memory Data Index container instance. For our use case it is not recommended and we should be using the data-index addon

I think the full property is `quarkus.kogito.devservices.enabled`

LightGuard replied:
Thank you

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+|`false`
+
+a|jbpm.devui.users.<user name>.groups
+a|Groups to which the provided user (specified in the name of the config) belongs.
```
this is used to configure which are the users and groups the dev-ui will be using in the task console to allow testing the process user tasks. <user name> refers to the name of the user, the value of this property should be a comma separated list with the groups for the specified user. 

This users will appear in the dev-ui's Task Inbox in a dropdown so the developer can switch between them and operate with tasks with the selected user identity.

A good example for this property would be:
```
jbpm.devui.users.jdoe.gropups=HR,task-operator
jbpm.devui.users.alice.groups=IT,task-operator,admin
```

gitgabrio replied:
Thanks @pefernan for explanation. 
But, what about -> **_groups_** <- themselves ? 
Are those defined elsewhere ?
Or, it is possible to write something like:

`jbpm.devui.users.goofy.groups=ny_own_unreferentiated_group,my_funny_group`

and then the groups will automatically be created/managed ?

pefernan replied:
@gitgabrio @LightGuard  don't have to be defined anywhere else, just in the property as you wrote.
This property is only for testing purposes. So when a developer can test the process execution in in quarkus:dev using the devui. This way, the developer can `impersonate` other users in the Task Inbox and test the user tasks with different users/roles.

LightGuard replied:
Thank you

pefernan commented:
@gitgabrio @LightGuard adding comments inline

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+a|false
+
+a|kogito.security.auth.impersonation.allowed-for-roles
+a|Comma-separated list of roles that allow identity impersonation when resolving the actual User Identity.
```
Traditionally User Tasks REST Endpoints used the user/group query params to generate the user identity (username and roles) that is performing a task operation (transtion, add attachment/comment, modify inputs/outputs...) 

This property works in conjunction with `kie.security.auth.enabled`. If auth is enabled, we are overriding the previous mechanism to use the info from the security context to generate the user identity instead of using the query params.

However, there are some times it is possible that a `super user` has to perform an operation `impersonating` another user (imagine an admin completing a blocked task in behalf of another user). This property allows to configure a comma separated list of roles that are allowed to impersonate other users. 
So, when performing a task operation (with auth enabled) and the user in the app security context belongs to any of the impersonator roles, in this case we will generate the user Identity based on the query params.


For example a good value would be:
```
kogito.security.auth.impersonation.allowed-for-roles=manager,task-admin
``` 
This property has no default value, so no default impersonator groups are defined, this is totally up to the customer to configure it with the roles they want.

gitgabrio replied:
Thanks @pefernan 
Same comment as https://github.com/IBM/bamoe-docs/pull/165#discussion_r1848181424

LightGuard replied:
@pefernan Excellent explanation! It will go into the next commit.

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,349 @@
+|`false`
+
+a|kogito.grpc.stream.timeout
+a|Time to wait for response from server when using streams.
```
This is SWF specific, do we have to document it too?

LightGuard replied:
Commenting out GRPC.

LightGuard commented:
> org.kie.kogito.addons.knative.eventing.sink.api-version -> possible values 
> org.kie.kogito.addons.knative.eventing.sink.kind -> possible values 

I would assume the values available are the ones at https://knative.dev/docs/eventing/reference/eventing-api/, which are exactly the same ones we're using. Unfortunately, the documentation over at knative is also lacking.

> kogito.grpc.stream.timeout -> time unit 

If this is indeed only for SonataFlow work, then we can drop it, as BAMOE does not support SonataFlow.

> kogito.devservices.shared -> quarkus only ? 

Probably, but as this is BAMOE documentation, that shouldn't be a problem.

> kogito.security.auth.impersonation.allowed-for-roles -> where are roles defined ? 

No clue, any of us security guys?

> multiple kogito.data-index urls

I have no idea here

> kogito.devservices.* -> are those available only for quarkus ?

Probably.

> there are some kogito.quarkus.events.* and kogito.events.* properties -> when are those available ?

Where are these?

pefernan commented:
@LightGuard Regarding your comment (https://github.com/IBM/bamoe-docs/pull/165#issuecomment-2495133775)

- `kogito.devservices.*` are quarkus specific, actually the full property names are `quarkus.kogito.devservices.*`. When defining a `@ConfigRoot` for Quarkus config, if no prefix is defined in the annotation it uses `quarkus`.

- multiple data-index urls: please look at https://github.com/IBM/bamoe-docs/pull/165#discussion_r1848090675

- `kogito.security.auth.impersonation.allowed-for-roles` -> in theory the roles specified here should be defined in the IDP where the runtime is configured to us (ej: keycloak).

- `kogito.events.*` this ones are related to the events addons, they are available for both quarkus and springboot.

- `kogito.quarkus.events.*` this ones are also related to to the eventings addons and mistakenly used for both quarkus and springboot (mistakenly by the explicit reference to quarkus in the property). This properties should be moved under `kogito.events.`

LightGuard commented:
@gitgabrio and @pefernan I believe I have addressed the comments from above.

gitgabrio commented:
Thanks @LightGuard 
we're almost there, I think 😄 

1. there are still "Predictions" reference (not part of BAMOE)
2. IMO it should be clarified that `kogito.quarkus.events.*` are not specific of quarkus environment
3. `When using IBM BAMOE, it is recommended to use the data-index addon instead of DevServices.`: it is not very clear to me: is it possible to have `data-index addon` AND DevServices ? Or, only one of them could be used, and DevServices should not be used with IBM BAMOE ?
4. `These roles vary on the application, and no default is given.` -> maybe could be rewritten with `Roles are defined inside the identity provider used by the current application` (or something similar): wdyt ?

@pefernan could you please double check, too, and provide clarification ? Thanks!

LightGuard commented:
@gitgabrio and @pefernan Please take another look, or just view the changes in the latest commit :)

gitgabrio commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,686 @@
+For example, an admin completing a blocked task on behalf of another user.
+These two configuration properties enable this "impersonating" to happen.
+
+Firstly, the `kogito.security.auth.enabled` proper must be set to true.
```
Ouch!
`proper`

LightGuard replied:
Got it, thanks!

gitgabrio reviewed (commented):
Thanks, Jason!
Alas, there is a typo.
Beside that, it looks good to me, but @pefernan (or any other expert in workflow area) should double-check properties related to it

pefernan commented on reference-guide/configuration.adoc:
```diff
@@ -1,4 +1,686 @@
+==== DevServices
+
+Kogito DevServices function only when a Kogito Quarkus application is in dev mode.
+Enabling DevServices starts an in-memory Data Index container.
```
```suggestion
When DevServices are enabled, if there's no Data-Index add-on present, Kogito will spin up a supporting in-memory Data-Index Container.
```

pefernan commented:
@LightGuard I'm ok with it, let's have the conflicts resolved and merge
